File: android_app/src/main/java/me/timeto/app/ui/NoteFormSheet.kt
package me.timeto.app.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.material.Icon
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.Delete
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.unit.dp
import me.timeto.app.*
import me.timeto.shared.db.NoteDb
import me.timeto.shared.vm.NoteFormSheetVm

@Composable
fun NoteFormSheet(
    layer: WrapperView.Layer,
    note: NoteDb?,
    onDelete: () -> Unit,
) {

    val (vm, state) = rememberVm(note) { NoteFormSheetVm(note) }

    VStack(
        modifier = Modifier
            .fillMaxHeight()
            .background(c.sheetBg)
            .imePadding(),
    ) {

        Sheet__HeaderView(
            title = state.headerTitle,
            scrollState = null,
            bgColor = c.sheetBg,
        )

        Column(
            modifier = Modifier
                .padding(bottom = H_PADDING)
                .weight(1f)
        ) {

            MyListView__ItemView(
                isFirst = true,
                isLast = true,
            ) {
                MyListView__ItemView__TextInputView(
                    placeholder = state.inputTextPlaceholder,
                    text = state.inputTextValue,
                    onTextChanged = { newText -> vm.setInputText(newText) },
                )
            }
        }

        Sheet__BottomViewDefault(
            primaryText = state.doneTitle,
            primaryAction = {
                vm.save {
                    layer.close()
                }
            },
            secondaryText = "Cancel",
            secondaryAction = {
                layer.close()
            },
            startContent = {
                val deleteFun = state.deleteFun
                if (deleteFun != null) {
                    Icon(
                        Icons.Rounded.Delete,
                        "Delete Note",
                        tint = c.red,
                        modifier = Modifier
                            .padding(start = H_PADDING - 0.dp)
                            .size(33.dp)
                            .clip(roundedShape)
                            .clickable {
                                deleteFun {
                                    layer.close()
                                    onDelete()
                                }
                            }
                            .padding(4.dp)
                    )
                }
            }
        )
    }
}

File: ios_app/iphone/NoteFormSheet.swift
import SwiftUI
import shared

struct NoteFormSheet: View {
    
    @EnvironmentObject private var nativeSheet: NativeSheet
    
    @Binding private var isPresented: Bool
    @State private var vm: NoteFormSheetVm
    private let onDelete: () -> Void
    
    init(
        isPresented: Binding<Bool>,
        note: NoteDb?,
        onDelete: @escaping () -> Void
    ) {
        _isPresented = isPresented
        _vm = State(initialValue: NoteFormSheetVm(note: note))
        self.onDelete = onDelete
    }
    
    var body: some View {
        
        VMView(vm: vm, stack: .VStack()) { state in
            
            Sheet__HeaderView(
                title: state.headerTitle,
                scrollToHeader: 0,
                bgColor: c.sheetBg
            )
            
            MyListView__ItemView(
                isFirst: true,
                isLast: true
            ) {
                
                MyListView__ItemView__TextInputView(
                    text: state.inputTextValue,
                    placeholder: state.inputTextPlaceholder,
                    isAutofocus: false,
                    onValueChanged: { newText in
                        vm.setInputText(newText: newText)
                    }
                )
            }
            .padding(.bottom, H_PADDING)
            
            Spacer()
            
            Sheet__BottomViewDefault(
                primaryText: state.doneTitle,
                primaryAction: {
                    vm.save {
                        isPresented = false
                    }
                },
                secondaryText: "Cancel",
                secondaryAction: {
                    isPresented = false
                },
                topContent: { EmptyView() },
                startContent: {
                    
                    ZStack {
                        
                        if let deleteFun = state.deleteFun {
                            
                            ZStack {
                                
                                Button(
                                    action: {
                                        deleteFun {
                                            // todo don't use to fast hide parent sheet
                                            isPresented = false
                                            onDelete()
                                            return KotlinUnit() // WTF?
                                        }
                                    },
                                    label: {
                                        ZStack {
                                            Image(systemName: "trash")
                                                .font(.system(size: 20, weight: .medium))
                                                .foregroundColor(c.red)
                                        }
                                        .frame(width: 34, height: 34)
                                    }
                                )
                            }
                            .padding(.leading, H_PADDING)
                        } else {
                            EmptyView()
                        }
                    }
                }
            )
        }
    }
}
